crates/js-sys/src/lib.rs

```diff
@@ -3641,6 +3641,31 @@ impl JsString {
     ) -> impl ExactSizeIterator<Item = u16> + DoubleEndedIterator<Item = u16> + 'a {
         (0..self.length()).map(move |i| self.char_code_at(i) as u16)
     }
+
+    /// If this string consists of a single UTF-8 codepoint, then this function
+    /// returns a Rust representation of that character without any allocations.
+    ///
+    /// If this JS value is not a valid UTF-8 or consists of more than a single
+    /// codepoint, then this returns `None`.
+    pub fn as_char(&self) -> Option<char> {
+        let len = self.length();
+
+        if len == 0 || len > 2 {
+            return None;
+        }
+
+        // This will be simplified when definitions are fixed:
+        // https://github.com/rustwasm/wasm-bindgen/issues/1362
+        let cp = self.code_point_at(0).as_f64().unwrap_throw() as u32;
+
+        let c = std::char::from_u32(cp)?;
+
+        if c.len_utf16() as u32 == len {
```

Comment:
Is this needed? My understanding is that `std::char::from_u32` already returns `None` for invalid characters (including surrogate pairs).

If you just want it for extra safety, couldn't it be a `debug_assert`?

Reply:
This is needed not for invalid characters, but to ensure this entire string is this character. E.g. without this check `"ab"` string would silently decode to just `'a'`.

Reply:
I've added a test to use such string to make it explicit.

Reply:
Ah, right, I hadn't considered that case.